src/precompiles/keccak256.rs

```diff
@@ -0,0 +1,194 @@
+use super::{precompile_abi_in_log, MemoryLocation, MemoryQuery, Precompile};
+use crate::{eravm_error::EraVmError, heaps::Heaps};
+use crypto_common::hazmat::SerializableState;
+use sha3::{Digest, Keccak256};
+use u256::U256;
+
+pub const KECCAK_RATE_BYTES: usize = 136;
+pub const MEMORY_READS_PER_CYCLE: usize = 6;
+pub const KECCAK_PRECOMPILE_BUFFER_SIZE: usize = MEMORY_READS_PER_CYCLE * 32;
+pub const MEMORY_WRITES_PER_CYCLE: usize = 1;
+
+pub struct ByteBuffer<const BUFFER_SIZE: usize> {
+    pub bytes: [u8; BUFFER_SIZE],
+    pub filled: usize,
+}
+
+impl<const BUFFER_SIZE: usize> ByteBuffer<BUFFER_SIZE> {
+    pub fn can_fill_bytes(&self, num_bytes: usize) -> bool {
+        self.filled + num_bytes <= BUFFER_SIZE
+    }
+
+    pub fn fill_with_bytes<const N: usize>(
+        &mut self,
+        input: &[u8; N],
+        offset: usize,
+        meaningful_bytes: usize,
+    ) {
+        assert!(self.filled + meaningful_bytes <= BUFFER_SIZE);
+        self.bytes[self.filled..(self.filled + meaningful_bytes)]
+            .copy_from_slice(&input[offset..(offset + meaningful_bytes)]);
+        self.filled += meaningful_bytes;
+    }
+
+    pub fn consume<const N: usize>(&mut self) -> [u8; N] {
+        assert!(N <= BUFFER_SIZE);
```

Comment:
Considering this is known at compile time, maybe assert this at compile time as well:
```suggestion
        const { assert!(N <= BUFFER_SIZE) };
```
I **think** this is legal to do and it forces build time evaluation of the assertion.